src/output/naive.rs

```diff
@@ -8,258 +8,214 @@
+
+        // load initial facts.
+        subset.insert(all_facts.outlives.into());
+        region_live_at.insert(Relation::from(all_facts.region_live_at.iter().map(|&(r, p)| ((r, p), ()))));
+        cfg_edge_p.insert(all_facts.cfg_edge.clone().into());
+
+        // .. and then start iterating rules!
+        while iteration1.changed() {
+
+            // remap fields to re-index by keys.
+            subset_r1p.from_map(&subset, |&(r1,r2,p)| ((r1,p),r2));
+            subset_r2p.from_map(&subset, |&(r1,r2,p)| ((r2,p),r1));
+            subset_p.from_map(&subset, |&(r1,r2,p)| (p,(r1,r2)));
+
+            // subset(R1, R2, P) :- outlives(R1, R2, P).
+            // Already loaded; outlives is static.
+
+            // subset(R1, R3, P) :-
+            //   subset(R1, R2, P),
+            //   subset(R2, R3, P).
+            subset.from_join(&subset_r2p, &subset_r1p, |&(_r2,p),&r1,&r3| (r1,r3,p));
+
+            // subset(R1, R2, Q) :-
+            //   subset(R1, R2, P),
+            //   cfg_edge(P, Q),
+            //   region_live_at(R1, Q),
+            //   region_live_at(R2, Q).
+
+            subset_1.from_join(&subset_p, &cfg_edge_p, |&_p, &(r1,r2), &q| ((r1,q),r2));
+            subset_2.from_join(&subset_1, &region_live_at, |&(r1,q),&r2,&()| ((r2,q),r1));
+            subset.from_join(&subset_2, &region_live_at, |&(r2,q),&r1,&()| (r1,r2,q));
+        }
+
+        subset_r1p.complete()
+    };
+
+    if dump_enabled {
+        println!("subset is complete: {} tuples, {:?}", subset.len(), subset_start.elapsed());
+
+        for ((r1, location), r2) in &subset.elements {
+            result
+                .subset
+                .entry(*location)
+                .or_insert(BTreeMap::new())
+                .entry(*r1)
+                .or_insert(BTreeSet::new())
+                .insert(*r2);
+            result.region_degrees.update_degrees(*r1, *r2, *location);
+        }
+    }
+
+    let requires_start = Instant::now();
+
+    // compute the requires rules, but indexed ((Region, Point), Loan) for the next iteration
+    let requires = {
+        // Create a new iteration context, ...
+        let mut iteration2 = Iteration::new();
+
+        // .. some variables, ..
+        let requires = iteration2.variable::<(Region, Loan, Point)>("requires");
+        requires.insert(all_facts.borrow_region.into());
+
+        // some indexes
+        let requires_rp = iteration2.variable::<((Region, Point), Loan)>("requires_rp");
+        let requires_bp = iteration2.variable::<((Loan, Point), Region)>("requires_bp");
+
+        let requires_1 = iteration2.variable::<(Point, (Loan, Region))>("requires_1");
+        let requires_2 = iteration2.variable::<((Region, Point), Loan)>("requires_2");
+
+        // since we're using subset mapped ((r, p), r) we can use it directly out of iteration 1
+        let subset_r1p = iteration2.variable::<((Region, Point), Region)>("subset_r1p");
+        subset_r1p.insert(subset);
```

Comment:
This "variable" is invariant, right? Do we really need to 'copy' the values from `subset` in here, or could we use `subset` directly?

(cc @frankmcsherry)

Reply:
Well, let me answer my own question: it won't build if we don't, because `from_join` wants a `Variable` and not a `Relation`: presumably that could be modified in datafrog? (Or is this `insert` very cheap anyway?)

Reply:
(I realize in this specific case we ought to collapse the iteration anyway.)

Reply:
(this is also relevant, it seems to the "redundantly computed index" entries we see below)

Reply:
Inserting a sorted relation should be pretty cheap. It will still do linear work, confirming that none of `subset` exists in the empty list of prior tuples, which we could totally optimize. But it shouldn't be too expensive either way.

More generally, there a bunch of cases where the difference between `Variable` and `Relation` and between `Key` and `(Key, ())` bite us a little. These can be picked off as appropriate; just some repetition in the library code in the interest of squeaking out some perf. :)

Reply:
Oh, my mistake. It will not even do the linear work, I think, on account of `self.stable` should be empty. I could be wrong about that, and we could work through exactly what happens when you install a relation. Not a great amount of care was used putting the wrappers in place. :)